src/main/scala/org/deepdive/inference/datastore/SQLInferenceDataStore.scala

```diff
@@ -459,173 +319,89 @@ trait SQLInferenceDataStore extends InferenceDataStore with Logging {
-        case RealArrayType(x) => 3
+      case Some(query) => {
+        log.info("Executing user supplied holdout query")
+        execute(query)
       }
-
-      val cardinality = dataType match {
-        case BooleanType => 2
-        case MultinomialType(x) => x.toInt
-        case RealNumberType => 2
-        case RealArrayType(x) => x.toInt
-      }
-
-      if(variableDataType == 2 || variableDataType == 3){
-
-      } else {
-        // This cannot be parsed in def randFunc for now.
-        // assign holdout - if not user-defined, randomly select from evidence variables of each variable table
-        calibrationSettings.holdoutQuery match {
-          case Some(s) =>
-          case None => execute(s"""
+      case None => {
+        log.info("There is no holdout query, will randomly generate holdout set")
+         // randomly assign variable holdout
+        schema.foreach { case(variable, dataType) =>
+          val Array(relation, column) = variable.split('.')
+          // This cannot be parsed in def randFunc for now.
+          // assign holdout - randomly select from evidence variables of each variable table
+          execute(s"""
             INSERT INTO ${VariablesHoldoutTable}
             SELECT id FROM ${relation}
             WHERE ${randomFunction} < ${calibrationSettings.holdoutFraction} AND ${column} IS NOT NULL;
             """)
         }
       }
+    }
 
+    // variable observation table
+    ds.dropAndCreateTable(VariablesObservationTable, "variable_id bigint primary key")
+    calibrationSettings.observationQuery match {
+      case Some(query) => {
+        log.info("Executing user supplied observation query")  
+        execute(query)
+      }
+      case None => {
+        log.info("There is no o query")
+      }
+    }
 
-      // Create a cardinality table for the variable
+  }
+  
+  // generate cardinality tables for variables in the schema
+  // cardinality tables is used to indicate the domains of the variables
+  def generateCardinalityTables(schema: Map[String, _ <: VariableDataType]) {
+    schema.foreach { case(variable, dataType) =>
+      val Array(relation, column) = variable.split('.')
       // note we use five-digit fixed-length representation here
       val cardinalityValues = dataType match {
         case BooleanType => "('00001')"
         case MultinomialType(x) => (0 to x-1).map (n => s"""('${"%05d".format(n)}')""").mkString(", ")
-        case RealNumberType => "('00001')"
-        case RealArrayType(x) => "('00001')"
       }
-      val cardinalityTableName = s"${relation}_${column}_cardinality"
+      val cardinalityTableName = InferenceNamespace.getCardinalityTableName(relation, column)
+      ds.dropAndCreateTable(cardinalityTableName, "cardinality text")
       execute(s"""
-        DROP TABLE IF EXISTS ${cardinalityTableName} CASCADE;
-        CREATE TABLE ${cardinalityTableName}(cardinality text);
         INSERT INTO ${cardinalityTableName} VALUES ${cardinalityValues};
```

Comment:
We might want to promulgate a new coding standard: no direct SQL queries without a function wrapper, into which parameters are passed.  The goal in doing that would be to enforce type checking by the compiler.  As it stands, if there's a type clash you don't find out until you try to run the query.  Not a huge deal for this commit, but probably a good idea in the future.